File: app/src/main/java/com/bearcave/passageplanning/routes/RouteManagerAdapter.java
package com.bearcave.passageplanning.routes;

import android.content.Context;
import android.util.SparseArray;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bearcave.passageplanning.R;
import com.bearcave.passageplanning.base.BaseManagerAdapter;
import com.bearcave.passageplanning.base.BaseManagerFragment;
import com.bearcave.passageplanning.data.database.tables.route.RouteDAO;
import com.bearcave.passageplanning.data.database.tables.waypoints.WaypointDAO;

import java.util.ArrayList;
import java.util.List;

import butterknife.ButterKnife;


public class RouteManagerAdapter extends BaseManagerAdapter {


    private ReadWaypoints waypointsDatabase;
    private SparseArray<WaypointDAO> waypoints;

    public RouteManagerAdapter(BaseManagerFragment parent, Context context) {
        super(parent, context);
        this.waypoints = new SparseArray<>();
        this.waypointsDatabase = (ReadWaypoints) context;

        for(WaypointDAO waypoint : waypointsDatabase.readAllWaypoints()){
            waypoints.put(waypoint.getId(), waypoint);
        }
    }


    private WaypointDAO getWaypointById(Integer id){
        return waypoints.get(id);
    }

    private WaypointDAO getWaypointFromList(int group, int child){
        return getWaypointById(getPassages().get(group).getWaypointsIds().get(child));
    }

    private ArrayList<RouteDAO> getPassages(){
        return getContainer();
    }


    @Override
    public int getChildrenCount(int groupPosition) {
        return getPassages().get(groupPosition).getWaypointsIds().size();
    }

    @Override
    public Object getChild(int groupPosition, int childPosition) {
        return getWaypointFromList(groupPosition, childPosition);
    }


    @Override
    public View getChildView(int groupPosition, int childPosition, boolean isLastChild, View convertView, ViewGroup parent) {
        View view = getInflater().inflate(R.layout.passage_manager_child_item, parent, false);
        WaypointDAO waypoint = getWaypointFromList(groupPosition, childPosition);

        TextView title = ButterKnife.findById(view, R.id.name);
        title.setText(waypoint.getName());

        ImageView up = ButterKnife.findById(view, R.id.up);
        ImageView down = ButterKnife.findById(view, R.id.down);

        return view;
    }
}

File: app/src/main/java/com/bearcave/passageplanning/routes/RouteManagerAdapter.kt
package com.bearcave.passageplanning.routes

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import butterknife.ButterKnife
import com.bearcave.passageplanning.R
import com.bearcave.passageplanning.base.BaseManagerAdapterWithWaypoints
import com.bearcave.passageplanning.routes.database.Route


class RouteManagerAdapter(parent: RouteManagerFragment, context: Context) : BaseManagerAdapterWithWaypoints<Route, Int>(parent, context) {


    init {
         addOption(context.getString(R.string.action_edit), { dao ->
             parent.openEditor(dao)
         })
    }


    private fun getWaypointFromList(group: Int, child: Int) = getWaypointById(
            container[group]
                    .waypointsIds[child]
        )

    override fun getChildrenCount(groupPosition: Int) = container[groupPosition].waypointsIds.size

    override fun getChild(groupPosition: Int, childPosition: Int) = getWaypointFromList(groupPosition, childPosition)

    override fun getChildView(groupPosition: Int, childPosition: Int, isLastChild: Boolean, convertView: View?, parent: ViewGroup?): View {
        val view = convertView ?: inflater.inflate(R.layout.passage_manager_child_item, parent, false)

        val (_, name) = getWaypointFromList(groupPosition, childPosition)

        val title = ButterKnife.findById<TextView>(view!!, R.id.name_placeholder)
        title.text = name

        val up = ButterKnife.findById<ImageView>(view, R.id.up)
        up.setOnClickListener {
            if (childPosition > 0) {
                val tmpWpts = container[groupPosition].waypointsIds
                val tmpId = tmpWpts[childPosition - 1]
                tmpWpts[childPosition - 1] = tmpWpts[childPosition]
                tmpWpts[childPosition] = tmpId
                notifyDataSetChanged()
                database.update(container[groupPosition])
            }
        }

        val down = ButterKnife.findById<ImageView>(view, R.id.down)
        down.setOnClickListener {
            if (childPosition < getChildrenCount(groupPosition) - 1) {
                val tmpWpts = container[groupPosition].waypointsIds
                val tmpId = tmpWpts[childPosition + 1]
                tmpWpts[childPosition + 1] = tmpWpts[childPosition]
                tmpWpts[childPosition] = tmpId
                notifyDataSetChanged()
                database.update(container[groupPosition])
            }
        }

        return view
    }
}
